components/eventtype/CustomInputTypeDialog.tsx

```diff
@@ -0,0 +1,61 @@
+import { PlusIcon } from "@heroicons/react/solid";
+import React, { ComponentProps, forwardRef, useImperativeHandle, useState } from "react";
+
+import { Dialog, DialogContent } from "@components/Dialog";
+
+import CustomInputTypeForm from "./CustomInputTypeForm";
+
+interface Props {
+  onSubmit: ComponentProps<typeof CustomInputTypeForm>["onSubmit"];
+  selectedCustomInput: ComponentProps<typeof CustomInputTypeForm>["selectedCustomInput"];
+}
+
+export type CustomInputTypeDialogRef = {
+  open: () => void;
+  close: () => void;
+};
+
+const CustomInputTypeDialog = forwardRef<CustomInputTypeDialogRef, Props>((props, ref) => {
+  const [open, setOpen] = useState(false);
+
+  useImperativeHandle(ref, () => ({
+    open: () => setOpen(true),
+    close: () => setOpen(false),
+  }));
```

Comment:
> * The parent doesn't need to know the state of the dialog, just being able to open or close it

It's nice to have modals linkable too -- see how "New event type" works on https://app.cal.com/event-types

> * By keeping the state in the child, we avoid re-rendering the whole parent when it's updated.

True, but this is not necessarily an actual problem and there's more declarative ways of solving that. Dan Abramov often talk about not optimizing React-stuff until it's an actual problem.

> * Also the less `useState`s we keep in a single component the better overall IMO.

Kinda agreed but this state has a good purpose.

Reply:
> It's nice to have modals linkable too -- see how "New event type" works on https://app.cal.com/event-types
How should we handle multiple dialogs, there's also the availability dialog pending.

Reply:
> Also removed this component and abstracted into `ImperativeDialog`. I think it's more legible and with less nesting overall.

Since you pass in `selectedCustomInput` anyway you remove any performance gain.

Don't get me wrong, I think it's great! However, React isn't imperative, it's declarative, and we should follow suit. There are a lot of perf tweaks you can do if that becomes a problem.

> How should we handle multiple dialogs, there's also the availability dialog pending.

I'll do an example - give me a sec!

Reply:
First example done - #863 - doing it without imperative dialog

Reply:
Second example - #864 - a bit more complex, I prob prefer #863